### DISCUSSION TITLE
xray内置DNS，有几个问题实在想不通

### DISCUSSION THREAD

**Discussion by Benjamin1919** [XTLS/Xray-core#2791] [Q&A]
用途：将xray作为本地DNS服务器 
环境：国内 客户端
在网上搜了许多关于xray/v2ray用作DNS服务器的配置，又重新研读了官方配置指南( https://xtls.github.io/config/dns.html )，自己写的配置如下：

```
{
  "dns": { //内置DNS配置
    "servers": [
       {
         "address": "119.29.29.29",
         "domains": [
           "geosite:cn" //国内域名用腾讯DNS解析
         ],
         "skipFallback": true
       },
       {
         "address": "8.8.8.8",
         "domains": [
           "geosite:geolocation-!cn" //国外域名用谷歌DNS解析
         ],
         "skipFallback": true
       },
       {
         "address": "1.1.1.1", //不匹配上述domain列表的用此服务器解析
         "skipFallback": false
       }
    ],
    "queryStrategy": "UseIP",
    "disableCache": false,
    "disableFallback": false,
    "disableFallbackIfMatch": false,
    "tag": "inner-dns"
  },
  "routing": { //路由规则配置
    "domainStrategy": "IPOnDemand",
    "domainMatcher": "hybrid",
    "rules": [
      {
        "type": "field",
        "network": "tcp,udp",
        "inboundTag": ["inner-dns"],
        "outboundTag": "direct" //内置DNS走直连出站
      },
      {
        "type": "field",
        "network": "tcp,udp",
        "inboundTag": ["dns-in"],
        "outboundTag": "dns-out"
      },
      {
        "type": "field",
        "network": "tcp,udp",
        "ip": ["8.8.8.8", "1.1.1.1"],
        "outboundTag": "proxy"
      }
    ]
  },
  "inbounds": [ //入站配置
    {
      "listen": "127.0.0.1",
      "port": 10053,
      "protocol": "dokodemo-door",
      "settings": {
        "address": "127.0.0.1",
        "port": 53,
        "network": "tcp,udp",
        "followRedirect": false
      },
      "tag": "dns-in"
    }
  ],
  "outbounds": [
    {
      "protocol": "dns",
      "settings": {
        "nonIPQuery": "skip"
      },
      "tag": "dns-out"
    },
    {
      "protocol": "freedom",
      "tag": "direct" //直连出站
    },
    {
      "protocol": "vless",
      "settings": {}, //代理设置省略
      "tag": "proxy" //代理出站
    }
  ]
}
```

大致意思是xray监听本地10053端口，DNS请求传入10053端口后由系统DNS服务器(127.0.0.1:53)转发给xray内置DNS进行解析，按照tag来写，xray的处理顺序如下：
"dns-in" → "dns-out" → "inner-dns" → "direct"

问题：

【1】我在路由规则里配置了8.8.8.8和1.1.1.1走代理出站，但也同时写了"inner-dns"直连出站，那么请问内置DNS向8.8.8.8或1.1.1.1发起的请求会走直连还是走代理？（假设没有开启透明代理）

【2】我看到有些人把入站"dns-in"的setting中的地址写成1.1.1.1:53，那么请问实际的DNS请求会先到1.1.1.1解析一次再转发给xray内置DNS进行解析吗？

【3】入站"dns-in"的setting中的地址和端口有什么玄机？是不是必须是某个DNS服务器？

【4】路由规则中如果没有写下面这一段，会发生什么？内置DNS还能正常解析吗？
```
      {
        "type": "field",
        "network": "tcp,udp",
        "inboundTag": ["inner-dns"],
        "outboundTag": "direct" //内置DNS走直连出站
      },
```
【5】outbounds中的"dns-out"如果写成下面这样，那么DNS最终是由1.1.1.1解析还是xray内置的DNS解析？
```
    {
      "protocol": "dns",
      "settings": {
        "address": "1.1.1.1",
        "port": 53,
        "nonIPQuery": "skip"
      },
      "tag": "dns-out"
    },
```

**Comment by kingwilliam** [comment]
Q1,Q4: 我在路由规则里配置了8.8.8.8和1.1.1.1走代理出站，但也同时写了"inner-dns"直连出站，那么请问内置DNS向8.8.8.8或1.1.1.1发起的请求会走直连还是走代理？（假设没有开启透明代理）
A1,A4: routing rules 是一行一行由上以下走的 那一个先配合 就即时用那个, 如你inner-dns比["8.8.8.8", "1.1.1.1"]更上层, 就已经用了direct

Q2,Q3,Q5: 我看到有些人把入站"dns-in"的setting中的地址写成1.1.1.1:53，那么请问实际的DNS请求会先到1.1.1.1解析一次再转发给xray内置DNS进行解析吗？
A2,A3,A5: 因xray dns server只会解释 A/AAAA 纪录, 如A/AAAA当然是你dns设定的 119.29, 1.1 8.8 但如果是问 mx或ns 这时就会交给 "dokodemo-door" 的预设值, 所以这应设定 119.29

A5(补充): 在处理 DNS 查询时，此出站协议会将 IP 查询（即 A 和 AAAA）转发给内置的 DNS 服务器。其它类型的查询流量将被转发至它们原本的目标地址 
https://xtls.github.io/config/outbounds/dns.html#outboundconfigurationobject
再简单点解说, 就是A/AAAA会走tag-out, 其他(如:mx/ns)就走1.1.1.1

==========

最後有一点要留意
"geosite:geolocation-!cn" //国外域名用谷歌DNS解析

"geosite:geolocation-!cn" 不等於 中国以外的domain 何解?
因geosite cn只包括大部份比较大的domain(如:BAT)
假如有一个国内小网站 pqrst.com, 这时就会用 1.1.1.1

我也不太确定, 希望有大神指正

**Comment by Benjamin1919** [comment]
> Q1,Q4: 我在路由规则里配置了8.8.8.8和1.1.1.1走代理出站，但也同时写了"inner-dns"直连出站，那么请问内置DNS向8.8.8.8或1.1.1.1发起的请求会走直连还是走代理？（假设没有开启透明代理） A1,A4: routing rules 是一行一行由上以下走的 那一个先配合 就即时用那个, 如你inner-dns比["8.8.8.8", "1.1.1.1"]更上层, 就已经用了direct
> 
> Q2,Q3,Q5: 我看到有些人把入站"dns-in"的setting中的地址写成1.1.1.1:53，那么请问实际的DNS请求会先到1.1.1.1解析一次再转发给xray内置DNS进行解析吗？ A2,A3,A5: 因xray dns server只会解释 A/AAAA 纪录, 如A/AAAA当然是你dns设定的 119.29, 1.1 8.8 但如果是问 mx或ns 这时就会交给 "dokodemo-door" 的预设值, 所以这应设定 119.29
> 
> A5(补充): 在处理 DNS 查询时，此出站协议会将 IP 查询（即 A 和 AAAA）转发给内置的 DNS 服务器。其它类型的查询流量将被转发至它们原本的目标地址 https://xtls.github.io/config/outbounds/dns.html#outboundconfigurationobject 再简单点解说, 就是A/AAAA会走tag-out, 其他(如:mx/ns)就走1.1.1.1
> 
> ==========
> 
> 最後有一点要留意 "geosite:geolocation-!cn" //国外域名用谷歌DNS解析
> 
> "geosite:geolocation-!cn" 不等於 中国以外的domain 何解? 因geosite cn只包括大部份比较大的domain(如:BAT) 假如有一个国内小网站 pqrst.com, 这时就会用 1.1.1.1
> 
> 我也不太确定, 希望有大神指正

非常感谢！另外再请教一下，如果DNS入站出站写成下面这样，那么非A和非AAAA记录是由223.5.5.5处理还是9.9.9.9处理呢？
A和AAAA记录是不是只会由xray内置DNS处理，完全不会经过223.5.5.5和9.9.9.9解析？
```
  "inbounds": [ 
    {
      "listen": "127.0.0.1",
      "port": 10053,
      "protocol": "dokodemo-door",
      "settings": {
        "address": "223.5.5.5",
        "port": 53,
        "network": "tcp,udp",
        "followRedirect": false
      },
      "tag": "dns-in"
    }
  ],
  "outbounds": [
    {
      "protocol": "dns",
      "settings": {
        "address": "9.9.9.9",
        "port": 53,
        "nonIPQuery": "skip"
      },
      "tag": "dns-out"
    }
  ]
```

**Reply by kingwilliam** [reply]
因我不是开发单位, 也未有空做单独测试, 只用过往一些经验一起研究.
如上述这个设定, 应该是走 9.9, 因dokodemo设定优先次序会比9.9低
